cartographer_grpc/mapping/trajectory_builder_stub.cc

```diff
@@ -16,38 +16,83 @@
 
 #include "cartographer_grpc/mapping/trajectory_builder_stub.h"
 
+#include "cartographer_grpc/proto/map_builder_service.pb.h"
 #include "glog/logging.h"
 
 namespace cartographer_grpc {
 namespace mapping {
 
 TrajectoryBuilderStub::TrajectoryBuilderStub(
-    std::shared_ptr<grpc::Channel> client_channel,
-    proto::MapBuilderService::Stub* stub)
-    : client_channel_(client_channel), stub_(stub) {}
+    std::shared_ptr<grpc::Channel> client_channel, const int trajectory_id)
+    : client_channel_(client_channel), trajectory_id_(trajectory_id) {
+  stub_ = proto::MapBuilderService::NewStub(client_channel_);
+  CHECK(stub_) << "Failed to create stub.";
+  rangefinder_writer_.client_writer = stub_->AddRangefinderData(
```

Comment:
We should find out what actually happens when you call e.g. `stub_->AddRangefinderData`. Does that open a TCP connection? Is there some network cost associated with holding a client writer even when you are not using it? If so we might want to create the writers lazily as we need them. I guess only in very few cases a bridge will actually want to send all 4 types of sensor data.

Reply:
Right. As discussed, we create each writer lazily before using it the first time.